lib/auth/auth_with_roles.go

```diff
@@ -4500,12 +4519,52 @@ func (a *ServerWithRoles) ReplaceRemoteLocks(ctx context.Context, clusterName st
 // channel if one is encountered. Otherwise the event channel is closed when the stream ends.
 // The event channel is not closed on error to prevent race conditions in downstream select statements.
 func (a *ServerWithRoles) StreamSessionEvents(ctx context.Context, sessionID session.ID, startIndex int64) (chan apievents.AuditEvent, chan error) {
-	if err := a.actionForKindSession(apidefaults.Namespace, types.VerbList, sessionID); err != nil {
+	createErrorChannel := func(err error) (chan apievents.AuditEvent, chan error) {
 		c, e := make(chan apievents.AuditEvent), make(chan error, 1)
 		e <- trace.Wrap(err)
 		return c, e
 	}
 
+	if err := a.actionForKindSession(apidefaults.Namespace, types.VerbList, sessionID); err != nil {
+		return createErrorChannel(err)
+	}
+
+	// StreamSessionEvents can be called internally, and when that happens we don't want to emit an event.
+	shouldEmitEvent := true
+
+	user, err := a.GetCurrentUser(context.Background())
+	if err != nil {
+		// If we can't fetch the user, it's probably an internal call (the Proxy role is missing from the backend)
+		// so we can skip emitting the event.
+
+		shouldEmitEvent = false
+	}
+
+	if shouldEmitEvent {
+		// However, if we do manage to fetch the user successfully, either for an internal call or a normal call, let's
+		// check the roles and make sure they don't include Auth or Proxy.
+		roles, err := types.NewTeleportRoles(user.GetRoles())
```

Comment:
This is invalid. `NewTeleportRoles` parses the built-in system roles. It will error out on any of the presets (access, editor, auditor) or other custom roles the user may have.

Reply:
Also, your `shouldEmitAuditEvent` doesn't really hold up to the name, because you still have to do the role check. I would try something like Tim does for the span exporter:

```
	// if forwarded on behalf of a Teleport service add its system roles
	_, isBuiltinRole := a.context.Identity.(BuiltinRole)
	shouldEmitAuditEvent := ~isBuiltinRole

```